cea/utilities/solar_equations.py

```diff
@@ -306,18 +307,94 @@ def optimal_angle_and_tilt(sensors_metadata_clean, latitude, solar_properties, m
         same as the roof. Sensors on flat roofs are all south facing.
     """
     # calculate panel tilt angle (B) for flat roofs (tilt < 5 degrees), slope roofs and walls.
-    optimal_angle_flat_deg = calc_optimal_angle(180, latitude,
+    input_angle_rad = radians(panel_tilt_angle)
+    sensors_metadata_clean['tilt_deg'] = np.vectorize(acos)(sensors_metadata_clean['Zdir'])  # surface tilt angle in rad
+    sensors_metadata_clean['tilt_deg'] = np.vectorize(degrees)(sensors_metadata_clean['tilt_deg'])  # surface tilt angle in degrees
+    sensors_metadata_clean['B_deg'] = np.where(sensors_metadata_clean['tilt_deg'] >= 5,
+                                               sensors_metadata_clean['tilt_deg'],
+                                               degrees(input_angle_rad))  # panel tilt angle in degrees
+
+    # calculate spacing and surface azimuth of the panels for flat roofs
+    module_length_m = panel_properties['module_length_m']
+    optimal_spacing_flat_m = calc_optimal_spacing(solar_properties, input_angle_rad, module_length_m)
+    sensors_metadata_clean['array_spacing_m'] = np.where(sensors_metadata_clean['tilt_deg'] >= 5, 0,
+                                                         optimal_spacing_flat_m)
+    sensors_metadata_clean['surface_azimuth_deg'] = np.vectorize(calc_surface_azimuth)(sensors_metadata_clean['Xdir'],
+                                                                                       sensors_metadata_clean['Ydir'],
+                                                                                       sensors_metadata_clean[
+                                                                                           'B_deg'])  # degrees
+
+    # calculate the surface area required to install one pv panel on flat roofs with defined tilt angle and array spacing
+    if panel_properties['type'] == 'PV':
+        module_width_m = module_length_m  # for PV
+    else:
+        module_width_m = panel_properties['module_area_m2'] / module_length_m  # for FP, ET
+    module_flat_surface_area_m2 = module_width_m * (sensors_metadata_clean.array_spacing_m / 2 +
+                                                    module_length_m * cos(input_angle_rad))
+    area_per_module_m2 = module_width_m * module_length_m
+
+    # calculate the pv/solar collector module area within the area of each sensor point
+    sensors_metadata_clean['area_installed_module_m2'] = np.where(
+        sensors_metadata_clean['tilt_deg'] >= 5, roof_coverage * sensors_metadata_clean.AREA_m2,
+        area_per_module_m2 * (roof_coverage * sensors_metadata_clean.AREA_m2 / module_flat_surface_area_m2))
```

Comment:
i think this is changing ALL of the coverage, including the walls

Reply:
again, great catch! fixed it